### PR TITLE
feat(install): add secure validators and run_cmd to validation.sh

### LINKED ISSUE #4795: Install2Fix - Enhance Input Validation Security (14/15)
## Problem
Validators do not sufficiently guard against command injection or path traversal.

## Scope
Enhance `validation.sh` with:
- validate_version_string_secure
- validate_path
- global `DRY_RUN` toggle for safe previews

## Approach
Whitelist regexes; reject metacharacters; refuse relative traversal.

## Dependencies
Blocked By:
- #4784
Blocks:
- #4796

## Files to Create/Modify
- [ ] scripts/install/common/validation.sh (modify)

## Phase
Phase 4: Security
Track: Security
Est. Hours: 6
PRs: 1

## Acceptance Criteria
- Functions reject unsafe inputs with clear messages.

## Testing Requirements
- Positive/negative strings and paths.

### File: scripts/install/common/validation.sh (additions)
```bash
# Secure validators
validate_version_string_secure() {
  local v="${1:-}"
  if [[ "$v" =~ ^[0-9]+(\.[0-9]+){1,2}$ ]]; then success "Version OK: $v"; return 0; fi
  error "Invalid version: $v (expected x.y or x.y.z)"; return 2
}
validate_path() {
  local p="${1:-}"
  case "$p" in
    /*) : ;;
    *) error "Path must be absolute: $p"; return 2 ;;
  esac
  if [[ "$p" == *".."* ]]; then error "Path traversal not allowed: $p"; return 2; fi
  success "Path OK: $p"
}
DRY_RUN="${DRY_RUN:-0}"
run_cmd() { if [ "$DRY_RUN" = "1" ]; then printf "[INFO] (dry-run) %s\n" "$*"; else eval "$@"; fi }
```

### PR DESCRIPTION
**Pull request by harsh04044**
**What kind of change does this PR introduce?**

Feature (security hardening for install scripts).

**Issue Number:**

Fixes #4795 

**Snapshots/Videos:**

N/A — changes are in shell scripts and tests; no UI.

**If relevant, did you update the documentation?**

No. This only adds helpers and tests under `scripts/install/common`. Install docs can be updated later if maintainers want to mention the new validators and `run_cmd`.

**Summary**

Install scripts use shared validation in `scripts/install/common/validation.sh`, but there was no strict version whitelist (e.g. x.y / x.y.z only), no path validation (absolute, no `..`), and no shared way to run commands with a dry-run. This PR adds:

- **validate_version_string_secure(v)** => Whitelist for simple version strings (x.y or x.y.z) to reduce risk of command injection from version-like input.
- **validate_path(p)** =>Ensures paths are absolute and rejects `..` to avoid path traversal.
- **DRY_RUN and run_cmd** =>Centralized command execution with optional dry-run; `run_cmd` uses `"$@"` (no `eval`) for safer invocation.

The file’s guard is updated to require a `success()` logger (already provided by `logging.sh` on macOS and inline on Linux), and the new helpers are documented and tested in `validation.test.sh` with positive/negative cases and existence checks. All 79 tests pass locally.

**Does this PR introduce a breaking change?**

No. New functions are additive. Existing callers that source `validation.sh` already define `success()` (Linux install script and macOS via `logging.sh`), so no changes are required elsewhere.

### CodeRabbit AI Review
- [x] I have reviewed and addressed all critical issues flagged by CodeRabbit AI
- [x] I have implemented or provided justification for each non-critical suggestion
- [x] I have documented my reasoning in the PR comments where CodeRabbit AI suggestions were not implemented

### Test Coverage
- [x] I have written tests for all new changes/features
- [x] I have verified that test coverage meets or exceeds 95%
- [x] I have run the test suite locally and all tests pass

**Other information**

Implementation follows existing patterns in `validation.sh` and `validation.test.sh` (block comments, return codes, regex-in-variable, and the same test layout). `run_cmd` is intentionally minimal and uses `"$@"` so callers pass the command and arguments as separate arguments rather than a single string.

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

## Summary by CodeRabbit

* **New Features**
  * Dry-run mode for safe command previews and a public success/feedback helper.

* **Improvements**
  * Safer command execution controls, stricter version and absolute-path validation, improved package metadata parsing, and standardized validation error messages with clearer retry guidance.

* **Tests**
  * Expanded coverage for validators, path checks, and command dry-run behavior; added extensive performance/timing tests for the event mutation.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by harsh04044** [comment]
The change in createEvent.metrics.test.ts is outside the validation hardening plan. It’s in this PR because CI was failing on that test (“10000 timers” infinite loop). Switching that one test to real timers fixes the failure without changing what’s being tested.

**Comment by harsh04044** [comment]
@coderabbitai full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved. Post your conclusion after review as an easily readable comment.

**Comment by harsh04044** [comment]
@coderabbitai full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved. Post your conclusion after review as an easily readable comment.

**Comment by harsh04044** [comment]
@coderabbitai full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved. Post your conclusion after review as an easily readable comment.

**Comment by harsh04044** [comment]
@coderabbitai full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved. Post your conclusion after review as an easily readable comment.

**Comment by harsh04044** [comment]
@coderabbitai   full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved. Post your conclusion after review as an easily readable comment.

**Comment by harsh04044** [comment]
@coderabbitai full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved. Post your conclusion after review as an easily readable comment.

**Comment by harsh04044** [comment]
@coderabbitai Post your review as comment

**Comment by harsh04044** [comment]
@palisadoes  review this

**Comment by harsh04044** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> `@harsh04044`, here's my comprehensive review of PR `#4951`:
> 
> # 🔒 Security Hardening Review - PR `#4951`
> ## Executive Summary: **APPROVED** ✅
> This PR successfully delivers production-ready security hardening for install-script validators, fully addressing issue `#4795` requirements.
> 
> ## 📦 What Was Delivered
> ### 🛡️ Security Validators (validation.sh)
> **1. validate_version_string_secure()**
> 
> * ✅ Enforces strict whitelist: `^[0-9]+(\.[0-9]+){1,2}$`
> * ✅ Accepts only: x.y or x.y.z formats
> * ✅ Blocks: prerelease tags, metadata, shell metacharacters
> * ✅ Prevents command injection via malicious version strings
> 
> **2. validate_path()**
> 
> * ✅ **Three-layer protection:**
>   
>   * Layer 1: Requires absolute paths (must start with `/`)
>   * Layer 2: Blocks path traversal (`..` patterns)
>   * Layer 3: Whitelist safe characters: `^/[a-zA-Z0-9/._-]*$`
> * ✅ Clear error messages for each failure case
> * ✅ Rejects shell metacharacters (`;`, `|`, `` ` ``, `$()`, etc.)
> 
> **3. run_cmd() with DRY_RUN**
> 
> * ✅ Safe execution using `"$@"` instead of dangerous `eval`
> * ✅ DRY_RUN toggle (default 0) for safe previews
> * ✅ Proper exit code propagation
> * ✅ Clean logging in dry-run mode
> 
> ### 🧪 Test Coverage (validation.test.sh)
> **79 comprehensive tests** covering:
> 
> * ✅ Guard enforcement (success() requirement)
> * ✅ Function existence checks
> * ✅ Positive validation cases
> * ✅ Negative validation cases
> * ✅ **Security-focused edge cases:**
>   
>   * Command injection attempts (`;`, `|`, `` ` ``, `$(...)`, redirects)
>   * Path traversal patterns
>   * Option injection attempts
>   * Empty strings and null bytes
> * ✅ DRY_RUN behavior verification
> * ✅ Helper function testing (parse_package_json, retry_command, handle_version_validation_error)
> 
> ## 🎯 Security Assessment
> Category	Rating	Details
> Command Injection Protection	⭐⭐⭐⭐⭐	Whitelist approach blocks all vectors
> Path Traversal Protection	⭐⭐⭐⭐⭐	Robust regex blocks `..` patterns
> Input Validation	⭐⭐⭐⭐⭐	Strict format enforcement
> Test Coverage	⭐⭐⭐⭐⭐	79 tests with adversarial inputs
> Safe Execution	⭐⭐⭐⭐⭐	Uses `"$@"`, avoids eval
> **🔒 No security vulnerabilities detected**
> 
> ## ⚠️ Issues & Notes
> ### 1. Out-of-Scope Change (createEvent.metrics.test.ts)
> **Your explanation:** "CI failing with '10000 timers' infinite loop"
> 
> **My assessment:** ✅ **Justified**
> 
> * Minimal change (4 lines)
> * Fixes genuine CI blocker
> * Doesn't alter test intent
> * **Recommendation:** Keep it—unblocking CI is pragmatic
> 
> ### 2. Docstring Coverage (50% vs 80% requirement)
> **Analysis:** Bash scripts use `#` comments, not traditional docstrings. Your implementation has **extensive documentation** (lines 1-42, 50-69, 73-132, etc.), but the tooling may not recognize bash comment format.
> 
> **This is a tooling/configuration issue, not a code quality issue.**
> 
> **Suggested fixes:**
> 
> * Check if the tool can be configured for shell scripts
> * Verify comment format matches tool expectations
> * Request waiver for bash scripts if tool is Python/JS-specific
> 
> ### 3. Minor: Root Path Edge Case
> `validate_path()` accepts `/` (root). While technically valid, root operations are risky.
> 
> **Recommendation:** Consider if you need root path support. If not, enforce minimum depth (`/[^/]+`). Not blocking—just flag for future consideration.
> 
> ## ✅ Strengths
> 1. **Enterprise-grade security:** Whitelist approach, defense in depth
> 2. **Comprehensive testing:** 79 tests with attack scenarios
> 3. **Excellent documentation:** Clear usage examples, security notes
> 4. **Backward compatible:** Additive changes, no breaking modifications
> 5. **Code consistency:** Follows existing patterns
> 6. **Clear error messages:** User-friendly guidance on failures
> 
> ## 🚀 Final Verdict
> ### **RECOMMENDATION: MERGE** (after resolving docstring coverage check)
> This PR delivers exactly what was specified in issue `#4795`:
> 
> * ✅ Secure version string validation
> * ✅ Path validation with traversal protection
> * ✅ Safe command execution with dry-run support
> * ✅ Comprehensive test suite
> * ✅ Clear error handling
> 
> **The security implementation is production-ready.** The only blocker is the docstring coverage metric, which is a tooling configuration issue unrelated to code quality.
> 
> ## 💡 Optional Post-Merge Enhancements
> 1. **Path length limits:** Add maximum path length (e.g., 4096) to prevent buffer overflow
> 2. **Version bounds:** Add maximum version number checks (e.g., reject `99999999.x`)
> 3. **Verbose mode:** Add optional debug logging showing which validation rule failed
> 4. **Root path:** Consider enforcing minimum path depth if root access isn't needed
> 
> ## 📊 Quality Metrics
> ```
> Code Quality:        ⭐⭐⭐⭐⭐ (5/5)
> Security:            ⭐⭐⭐⭐⭐ (5/5)
> Test Coverage:       ⭐⭐⭐⭐⭐ (5/5)
> Documentation:       ⭐⭐⭐⭐⭐ (5/5)
> Backward Compat:     ⭐⭐⭐⭐⭐ (5/5)
> ```
> 
> **Outstanding work on the security hardening!** 🎉🔒
> 
> _Review based on: validation.sh (+127/-25), validation.test.sh (+201/-4), createEvent.metrics.test.ts (+39/-6)_

@coderabbitai i have addressed docstring coverage verify it

**Comment by palisadoes** [comment]
@coderabbitai full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved. 

Evaluate whether the goals of the issue linked in the first PR comment are met.

Post your conclusion as an easily readable comment.

**Comment by palisadoes** [comment]
@coderabbitai What is your conclusion based on this request? Post your response as an easily readable comment.
- https://github.com/PalisadoesFoundation/talawa-api/pull/4951#issuecomment-3823118304

**Comment by palisadoes** [comment]
Please fix the conflicting files

**Comment by harsh04044** [comment]
@coderabbitai full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved.

Evaluate whether the goals of the issue linked in the first PR comment are met.

Post your conclusion as an easily readable comment.

**Comment by palisadoes** [comment]
@coderabbitai full review and reevaluate all your previous comments that may have been dismissed, ignored or resolved. 

Evaluate whether the goals of the issue linked in the first PR comment are met.

Post your conclusion as an easily readable comment.

**Comment by palisadoes** [comment]
@coderabbitai What is your conclusion based on this request? Post your response as an easily readable comment.
- https://github.com/PalisadoesFoundation/talawa-api/pull/4951#issuecomment-3824611109

File: scripts/install/common/validation.test.sh [hunk 1]
```diff
@@ -172,6 +203,176 @@ else
...
+test_invalid_path '/tmp/$(rm -rf /)' "command substitution"
+test_invalid_path '/tmp/foo&' "background ampersand"
+test_invalid_path '/tmp/foo>out' "redirection"
+test_invalid_path '/tmp/foo<in' "input redirection"
+test_invalid_path '/tmp/foo*' "glob asterisk"
+test_invalid_path '/tmp/foo?' "glob question mark"
+test_invalid_path $'/tmp/foo\'out' "single quote"
+test_invalid_path '/tmp/foo"out' "double quote"
+test_invalid_path '/tmp/foo\out' "backslash"
+test_invalid_path '/tmp/foo out' "space"
```
[the remark this answers is not in the document]
Comment (palisadoes) [reply scripts/install/common/validation.test.sh:329]: @coderabbitai Create an issue that is unassigned to resolve this concern.
